Remove use of g_application from CApplication

garbear opened the pull request:
This removes use of g_application and CServiceBroker from CApplication.

Instead of g_application, local variables are referenced directly. Instead of CServiceBroker, access occurs through m_ServiceManager.

I've beefed up the PR description to explain a little more:

## Background
In 2016, fernet submitted #9389 which addressed a crash on exit I couldn't fix. It adds two classes:

* CServiceManager
* CServiceBroker

CServiceManager is a temporary solution used as a tool to change a global into a local variable. In CServiceManager, the startup/shutdown order is controlled. Notice that shutdown happens exactly in the opposite order than startup.

CServiceBroker is used to access the previously-global services. The point of a second class is "separation of concerns" - storage of variables is separated from retrieval.

Look at ServiceBroker.h, it does not include anything. So clients won’t depend on anything by including ServiceBroker. Further clients are not interested in how services are managed or the order in which they get launched.

## Credits

Thanks to fernetmenta for the general architecture and some verbiage above.

garbear committed:
Remove use of g_application from CApplication

FernetMenta commented:
this makes it harder to move CServiceBroker out of CApplication. CApp is already a monster and should not grow. Every client that includes CApplication.h does also get internals of ServiceManager.

garbear commented:
You're right, the CServiceBroker commit should be dropped. I won't merge this PR in its current state.

garbear commented:
K, dropped the service manager commit. By your admission, ServiceBroker.cpp shouldn't include any headers, so is the next step to factor out the dynamic_cast?

FernetMenta commented:
ServiceBroker.h (not cpp) should not include headers.

FernetMenta commented:
> so is the next step to factor out the dynamic_cast?

RenderSystem is solely onwed by WinSystem.  GetWinSystem->GetRenderSystem() would be better than the dynamic_cast

garbear committed:
Move RenderSystem access to WinSystem

RenderSystem is solely onwed by WinSystem. GetWinSystem->GetRenderSystem()
would be better than the dynamic_cast in ServiceBroker.cpp.

garbear commented:
@FernetMenta I removed the dynamic_cast using the strategy you recommended. Is there anything else I can do in the scope of improving our service management architecture?

FernetMenta commented on xbmc/ServiceManager.cpp:
```diff
@@ -421,6 +422,11 @@ CWinSystemBase &CServiceManager::GetWinSystem()
   return *m_winSystem.get();
 }
 
+CRenderSystemBase *CServiceManager::GetRenderSystem()
```
this method is wrong. ServiceManager does not manage RenderSystem, it does not even know that it exists. RenderSystem is owned by WinSystem.

garbear replied:
Why does ServiceBroker know about RenderSystem then?

FernetMenta replied:
it was intermediate. clients can call GetWinSystem()->GetRenderSystem()

garbear replied:
@FernetMenta updated

garbear commented:
jenkins build this please (might have missed a windowing file)

FernetMenta reviewed (approved):
thanks
